File: алгоритмы/divide_and_conquer(разделяй_и_властвуй)/Sort_Сортировка/sort_merge.rb
arr = [3, 22, 3, -1, 11, 12, 4, 1, 8, 14]

# ! Сортировка слиянием
def sort_merge(arr)
  len = arr.size

  if len > 1
    # округлим до меньшего числа
    mid = (len / 2).floor

    # сделаем копии массивов в указанных диапазонах
    left_arr = arr[0...mid]
    right_arr = arr[mid..(len - 1)]

    # применим рекурсию для разбивания массива до единичного отсортированного элемента в нем
    sort_merge(left_arr)
    sort_merge(right_arr)

    l = r = 0

    len.times do |k|
      if left_arr[l].nil?
        arr[k] = right_arr[r]
        r += 1
      elsif right_arr[r].nil?
        arr[k] = left_arr[l]
        l += 1
      elsif left_arr[l] <= right_arr[r]
        arr[k] = left_arr[l]
        l += 1
      else
        arr[k] = right_arr[r]
        r += 1
      end
    end
  end

  arr
end

p arr
p sort_merge(arr)

# * =========================================================================================

arr2 = [3, 22, 3, -1, 11, 12, 4, 1, 8, 14]
# ! Сортировка слиянием вариант 2
def sort_merge2(arr)
  len = arr.size

  if len > 1
    mid = (len / 2).floor

    left_arr = arr[0...mid]
    right_arr = arr[mid..(len - 1)]

    sort_merge2(left_arr)
    sort_merge2(right_arr)

    l = r = k = 0
    size_left, size_right = left_arr.size, right_arr.size

    while l < size_left && r < size_right
      if left_arr[l] <= right_arr[r]
        arr[k] = left_arr[l]
        l += 1
      else
        arr[k] = right_arr[r]
        r += 1
      end
      
      k += 1
    end

    # добиваем если остались
    while l < size_left
      arr[k] = left_arr[l]
      l += 1
      k += 1
    end

    # добиваем если остались
    while r < size_right
      arr[k] = right_arr[r]
      r += 1
      k += 1
    end
  end

  arr
end

p arr2
p sort_merge2(arr2)

File: алгоритмы/сортировки/divide_and_conquer(разделяй_и_властвуй)/sort_merge.js
let arr = [3, 22, 3, -1, 11, 12, 4, 1, 8, 14];

// ! Сортировка слиянием
function sortMerge(arr) {
  let len = arr.length;

  if (len > 1) {
    // высчитаем среднюю длину округлив к минимальному значению
    let mid = Math.floor(len / 2);

    // срежем часть массива по индексу от 0 включительно и до mid не включительно (вернет новый массив)
    let leftArr = arr.slice(0, mid);
    // срежем от mid включительно и до конца (вернет новый массив)
    let rightArr = arr.slice(mid);

    // после разделения на две части продолжим дробление массива до тех пор пока не останется один элемент (рекурсия)
    sortMerge(leftArr);
    sortMerge(rightArr);

    let l = (r = k = 0); // ! 'use strict' не будет работать если установлен в документе;

    while (l < leftArr.length && r < rightArr.length) {
      if (leftArr[l] <= rightArr[r]) {
        arr[k] = leftArr[l];
        l += 1;
      } else {
        arr[k] = rightArr[r];
        r += 1;
      }

      k += 1;
    }

    // добить при необходимости
    while (l < leftArr.length) {
      arr[k] = leftArr[l];
      l += 1;
      k += 1;
    }

    // добить при необходимости
    while (r < rightArr.length) {
      arr[k] = rightArr[r];
      r += 1;
      k += 1;
    }
  }

  return arr;
}

console.log(sortMerge(arr));

// * =========================================================================================

let arr2 = [3, 22, 3, -1, 11, 12, 4, 1, 8, 14];

// ! Сортировка слиянием
function sortMerge(arr) {
  let len = arr.length;

  if (len > 1) {
    let mid = Math.floor(len / 2);

    let leftArr = arr.slice(0, mid);

    let rightArr = arr.slice(mid);

    sortMerge(leftArr);
    sortMerge(rightArr);

    let l = (r = k = 0); // ! 'use strict' не будет работать если установлен в документе;

    while (k < len) {
      if (leftArr[l] <= rightArr[r]) {
        arr[k] = leftArr[l];
        l += 1;
      } else if (leftArr[l] > rightArr[r]) {
        arr[k] = rightArr[r];
        r += 1;

        // leftArr[l] == undefined //  leftArr[l] == false
      } else if (!leftArr[l]) {
        arr[k] = rightArr[r];
        r += 1;

        // rightArr[r] == undefined остается только один вариант по этому можно и else указать
      } else {
        arr[k] = leftArr[l];
        l += 1;
      }

      k += 1;
    }
  }

  return arr;
}

console.log(sortMerge(arr2));
